controllers/csm_controller.go

```diff
@@ -1524,12 +1529,68 @@ func (r *ContainerStorageModuleReconciler) GetK8sClient() kubernetes.Interface {
 	return r.K8sClient
 }
 
-func (r *ContainerStorageModuleReconciler) GetMatchingNodes(ctx context.Context, labelKey string, labelValue string) (*corev1.NodeList, error) {
+func (r *ContainerStorageModuleReconciler) GetMatchingNodes(ctx context.Context, labelKey string) (*corev1.NodeList, error) {
 	nodeList := &corev1.NodeList{}
 	opts := []client.ListOption{
-		client.MatchingLabels{labelKey: labelValue},
+		client.HasLabels{labelKey},
 	}
 	err := r.List(ctx, nodeList, opts...)
 
 	return nodeList, err
 }
+
+// Invalid zoning configurations:
+// 1. Some arrays have a zone label A, but one of the arrays has zone label B, or no label
+// 2. Array secret has zone label A, but one of the nodes has zone label B, or no label
+// 3. Some of the nodes have a zone label, but arrays have no zone labels.
+// 4. All of the arrays have a zone label, but nodes have no zone label.
+func (r *ContainerStorageModuleReconciler) ZoneValidation(ctx context.Context, cr *csmv1.ContainerStorageModule, namespace string) error {
+	log := logger.GetLogger(ctx)
+	nodeList := &corev1.NodeList{}
+
+	err := r.Client.List(ctx, nodeList)
+	if err != nil {
+		return fmt.Errorf("error listing nodes %v", err)
+	}
+	numberOfNodes := len(nodeList.Items)
+
+	zones, err := drivers.ExtractZones(ctx, cr, r.Client, namespace)
+	if err != nil {
+		return fmt.Errorf("ZoneValidation failed with error: %v", err)
+	}
+
+	// Make sure all zone keys are the same, if not fail
+	var firstKeyLabel string
+	for key := range zones {
+		if firstKeyLabel == "" {
+			firstKeyLabel = key
+			continue
+		}
+		if key != firstKeyLabel {
+			return fmt.Errorf("detected mismatched Keys in zones")
+		}
+	}
+	// TBD: also check if there any arrays specified in the secret that do not have
+	// a zone label. this constitutes a zone configuration failure.
+	// Determine how many arrays are configured and match that with the len(map)
+
+	// The labelKey, GetMatchingNodes with labelKey, verify that the list
+	// returned is the same size as the current number of nodes on the cluster
+	// i.e. all nodes must have this zoning label key. if the number of nodes
+	// returned does not match the number of nodes on the cluster, then fail,
+	// otherwise return true
+	if firstKeyLabel != "" {
```

Comment:
we should probably double check w/ @tdawe but I think its ok for nodes not to have a label -> this means customer doesn't want driver to connect to any array on this node

Reply:
that's different from what we understood.

Either all nodes in the cluster (except master) have zone labels, or none of them do. there's no in-between, at least that's my understanding.

Reply:
The customer may only want the DaemonSet to be running on certain nodes (see option for this with the nodeSelector: https://github.com/dell/csm-operator/blob/main/samples/storage_csm_powerflex_v2130.yaml#L167). Therefore, the zoning label is not required to be on all nodes in the k8s cluster.

Reply:
PR closed and code removed as part of another PR.